category/traversable/derive.lean

```diff
@@ -0,0 +1,212 @@
+            _inst ← mk_instance cl,
+            f' ← nested_traverse t.app_arg,
+            mk_mapp ``traversable.traverse [t.app_fn,_inst,none,none,none,none,f']
+          else fail format!"type {t} is not traversable with respect to variable {v}")
+
+meta def traverse_field (n : name) (_inst cl f v e : expr) : tactic expr :=
+do t ← infer_type e >>= whnf,
+   if t.get_app_fn.const_name = n
+   then fail "recursive types not supported"
+   else if v.occurs t
+   then do f' ← nested_traverse f v t,
+           pure $ f' e
+   else
+         (is_def_eq t.app_fn cl >> to_expr ``(comp.mk %%e))
+     <|> to_expr ``(@pure _ (%%_inst).to_has_pure _ (ulift.up %%e))
+
+meta def seq_apply_constructor : list expr → expr → tactic expr
+| (x :: xs) e := to_expr ``(%%e <*> %%x) >>= seq_apply_constructor xs
+| [] e := return e
+
+meta def fill_implicit_arg' : expr → expr → tactic expr
+| f (expr.pi n bi d b) :=
+if b.has_var then
+do v ← mk_meta_var d,
+   fill_implicit_arg' (expr.app f v) (b.instantiate_var v)
+else return f
+| e _ := return e
+
+meta def fill_implicit_arg (n : name) : tactic expr :=
+do c ← mk_const n,
+   t ← infer_type c,
+   fill_implicit_arg' c t
+
+meta def mk_down (e : expr) : tactic expr :=
+to_expr ``(ulift.down %%e) <|> pure e
+
+meta def map_constructor (c n : name) (f v : expr) (args : list expr) : tactic unit :=
+do g ← target,
+   args' ← mmap (map_field n g.app_fn f v) args,
+   constr ← fill_implicit_arg c,
+   let r := constr.mk_app args',
+   () <$ tactic.exact r
+
+meta def mk_map (type : name) (cs : list name) := do
+   `[intros α β f x],
+   reset_instance_cache,
+   x ← get_local `x,
+   xs ← tactic.induction x,
+   f ← get_local `f,
+   α ← get_local `α,
+   β ← get_local `β,
+   () <$ mzip_with'
+      (λ (c : name) (x : name × list expr × list (name × expr)),
+      solve1 (map_constructor c type f α x.2.1))
+      cs xs
+
+meta def traverse_constructor (c n : name) (_inst f v : expr) (args : list expr) : tactic unit :=
+do g ← target,
+   args' ← mmap (traverse_field n _inst g.app_fn f v) args,
+   constr ← fill_implicit_arg c,
+   v ← mk_mvar,
+   constr' ← to_expr ``(@pure %%(g.app_fn) (%%_inst).to_has_pure _ %%v),
+   r ← seq_apply_constructor args' constr',
+   gs ← get_goals,
+   set_goals [v],
+   vs ← tactic.intros >>= mmap mk_down,
+   tactic.exact (constr.mk_app vs),
+   done,
+   set_goals gs,
+   () <$ tactic.exact r
+
+meta def mk_traverse (type : name) (cs : list name) := do
+   `[intros m _inst α β f x],
+   reset_instance_cache,
+   x ← get_local `x,
+   xs ← tactic.induction x,
+   f ← get_local `f,
+   _inst ← get_local `_inst,
+   α ← get_local `α,
+   β ← get_local `β,
+   m ← get_local `m,
+   () <$ mzip_with'
+      (λ (c : name) (x : name × list expr × list (name × expr)),
+      solve1 (traverse_constructor c type _inst f α x.2.1))
+      cs xs
+
+open applicative
+
+meta def derive_functor : tactic unit :=
+do `(functor %%f) ← target | failed,
```

Comment:
Is `| failed` required here? I thought that was the default behavior if you don't specify